### DISCUSSION TITLE
repeater checkbox response very slow

### DISCUSSION THREAD

**Discussion by samg666** [Budibase/budibase#6548] [Help]
Hello,

I have put checkbox on my repeater, when the checkbox checked(On Change) then it will store selected row's id to `app state` for bulk update purpose. The problem is, the checkbox response very slow. Below is my checkbox on change code. The code is not complex, but I am not sure why it's such slow. Is there any better way store selected rows ids(I am using MySQL custom query, the tables componemt's `Allow row selection` didn't support custom SQL )?

<img width="1384" alt="image" src="https://user-images.githubusercontent.com/29649651/177172597-7cfe74ca-507a-41c7-9808-c7b61fc3d118.png">

```
let arry = $("State.IPList_SelectedIPs") ? $("State.IPList_SelectedIPs") : []
arry.push($("RPT_IPs.SelectIPsView.id"))
return arry
```

![multi-select](https://user-images.githubusercontent.com/29649651/177170621-08ee7c79-bc7e-44dc-9336-8f4cf65f0b99.gif)

**Comment by samg666** [comment]
Here is Chrome performance, it seems the CheckBox On Change event takes long time to execute, but the On Change code seems not complex, not sure why it takes such long time.

<img width="2385" alt="image" src="https://user-images.githubusercontent.com/29649651/177178133-1ec98a08-0da9-4ed3-bc25-f14ea05f547d.png">

**Comment by PClmnt** [comment]
Hi @ggs331, 

It seems like you are only ever adding data to the App State with your code, i.e, you have no code in place for whenever the checkbox gets de-selected. However, does this slowness happen even if the page has just loaded and one checkbox has been selected, or does it get slowly worse the more checkboxes you have selected?

**Reply by samg666** [reply]
Hi @PClmnt 

I do have de-select and remove code, I have removed it just to keep the code more simple for testing purpose, but the click still lagging. The 1st click was slow, then 2nd... even slower.

**Comment by aptkingston** [comment]
Hey @samg666. I recreated this myself and it performs very well, so I'm not sure why yours is so slow. Could I see your component tree? I wonder if maybe there are nested repeaters or just too many components on the page. Updating state is one of the more intensive operations as it forces an update of every single component on the page.

For reference, here is my example of doing this which performs very well:

https://user-images.githubusercontent.com/9075550/188625215-28580e16-b844-446a-b90f-be3703c1a2a2.mp4

And here's how my component tree looks for this example:
![image](https://user-images.githubusercontent.com/9075550/188625323-0c6e319b-9a05-4658-b4a3-139c34fbd31c.png)

**Reply by samg666** [reply]
@aptkingston Thank you very much for your reply. Below are my details. I have this issue for long time and still not find a way to fix it. I know the table component can support multi-select well, but as I used many custom SQL queries that tables cannot support multi-select yet. 

I will try re-structure as your component tree, and it would be really helpful if you can share your checkbox code?

### Here is my component tree
<img width="452" alt="image" src="https://user-images.githubusercontent.com/29649651/188930042-b5a8e894-e1ff-4812-8e35-ff8f66d138f6.png">

### below is my checkbox `Update State > Set value` code
```
let a=$("State.ROW_SELECT");
if (Array.isArray(a)) {
	if ($("Field Value") == true) {
		a.push($("RPT_Servers.SelectServers.id"));
		return a;
	} else {
		var index = a.indexOf($("RPT_Servers.SelectServers.id"));
		if (index !== -1){
			a.splice(index, 1);
			return a;
		}
	}
} else {
	if ($("Field Value") == true) {
		return [$("RPT_Servers.SelectServers.id")];
	} else {
		return [];
	}
}
```

### Below is the demo, click on the checkbox is laggy 
![iShot_2022-09-07_23 39 43](https://user-images.githubusercontent.com/29649651/188930489-d322ec31-b055-43c5-8fb2-4d5db3044e54.gif)

**Comment by samg666** [comment]
Hi @aptkingston I have created a page from scratch with same structure as yours, yes, it works very good, no laggy. 

Then back to  my existing laggy page, I removed all of my repeater cell(actually paragraph)'s text binding or conditional scripts, then checkbox click become response very fast. 

I think it's maybe due to the checkbox click not just triggered checkbox's onChange scripts, but also all other scripts inside the repeater, the more binding/conditional scripts on my repeater cells, the more laggy.

Anyway, I think I found the root cause. My workaround is remove all binding and conditional scripts, just keep the repeater cells simply bind to repeater fields directly without any scripts.

### repeater cells with text binding or conditional scripts, then checkbox click response laggy
![repeater_cells_with_scripts](https://user-images.githubusercontent.com/29649651/189059506-49e41389-0f5f-424e-9e24-7dd27cd2adb0.gif)

### repeater cells *without* any text binding or conditional scripts, then checkbox click response very fast
![repeater_cells_without_scripts](https://user-images.githubusercontent.com/29649651/189059702-bc4bc629-65d9-45ed-8319-05064a191ebe.gif)
